compiler/rustc_middle/src/mir/mono.rs

```diff
@@ -273,13 +276,38 @@ pub struct MonoItemData {
     /// `GloballyShared` maps to `false` and `LocalCopy` maps to `true`.
     pub inlined: bool,
 
-    pub linkage: Linkage,
+    pub linkage: LinkageKind,
     pub visibility: Visibility,
 
     /// A cached copy of the result of `MonoItem::size_estimate`.
     pub size_estimate: usize,
 }
 
+/// Specifies the linkage type for a `MonoItem`, and how we know that linkage to use
+#[derive(Copy, Clone, PartialEq, Debug, TyEncodable, TyDecodable, HashStable)]
+pub enum LinkageKind {
+    /// The linkage was specified explicitly (e.g. using #[linkage = "..."])
+    Explicit(Linkage),
+    /// Assume the symbol may be used from other CGUs, a safe default
+    ImplicitExternal,
+    /// We did not find any uses from other CGUs, so it's fine to make this internal
+    ImplicitInternal,
+}
```

Comment:
can I get a vibe-check on this type (introduced in https://github.com/rust-lang/rust/pull/128004/commits/6b8141d1165d27b6e714fed60b0bf96db2e76216). This keeps a bit more information about how we decide what linkage to use around, so the backend can be flexible and still use external linkage when the linkage is undefined (and could safely default to internal). That is required for our naked function approach to connect the extern fn declaration and the global asm body.

I'll probably pull this out into its own PR if this approach seems promising, because this one is getting large